app/uk/gov/hmrc/gform/obligation/ObligationService.scala

```diff
@@ -17,30 +17,108 @@
 package uk.gov.hmrc.gform.obligation
 import java.text.SimpleDateFormat
 
+import uk.gov.hmrc.gform.auth.AuthService
+import uk.gov.hmrc.gform.auth.models.MaterialisedRetrievals
+import uk.gov.hmrc.gform.controllers.helpers.FormDataHelpers
 import uk.gov.hmrc.gform.gformbackend.GformConnector
 import uk.gov.hmrc.gform.sharedmodel._
+import uk.gov.hmrc.gform.sharedmodel.form.{ Form, FormId, UserData }
 import uk.gov.hmrc.gform.sharedmodel.formtemplate._
 import uk.gov.hmrc.http.HeaderCarrier
-import scala.concurrent.ExecutionContext
+
+import scala.concurrent.{ ExecutionContext, Future }
 
 class ObligationService(gformConnector: GformConnector) {
 
   val stringToDate = new SimpleDateFormat("yyyy-MM-dd")
 
-  def lookupObligationsMultiple(formTemplate: FormTemplate)(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
+  def lookupObligationsMultiple(
+    formTemplate: FormTemplate,
+    authService: AuthService,
+    retrievals: MaterialisedRetrievals,
+    form: Form)(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
+    val hmrcTaxPeriodIdentifiers = formTemplate.expandFormTemplateFull.allFCs.collect {
+      case IsHmrcTaxPeriod(el) => el
+    }
+    for {
+      listOfIdNumbers <- Future.traverse(hmrcTaxPeriodIdentifiers)(
+                          i =>
+                            authService.evaluateSubmissionReference(
+                              i.idNumber,
+                              retrievals,
+                              formTemplate,
+                              FormDataHelpers.formDataMap(form.formData),
+                              form.envelopeId))
+      replaceIdNumber = listOfIdNumbers.zip(hmrcTaxPeriodIdentifiers)
+      output <- gformConnector
+                 .getAllTaxPeriods(replaceIdNumber.map(a =>
+                   HmrcTaxPeriod(a._2.idType, TextExpression(Constant(a._1)), a._2.regimeType)))
+                 .map(i =>
+                   i.flatMap(j => j.obligation.obligations.flatMap(h => makeAllInfoList(j.id, h.obligationDetails))))
+    } yield output
+  }
+
+  def makeAllInfoList(id: HmrcTaxPeriod, obligation: List[ObligationDetail]) =
+    obligation.map(i =>
+      TaxPeriodInformation(id, i.inboundCorrespondenceFromDate, i.inboundCorrespondenceToDate, i.periodKey))
+
+  def lookupIfPossible(
+    form: Form,
+    formTemplate: FormTemplate,
+    authService: AuthService,
+    retrievals: MaterialisedRetrievals)(implicit hc: HeaderCarrier, ec: ExecutionContext) = {
     val hmrcTaxPeriodIdentifiers = formTemplate.expandFormTemplateFull.allFCs.collect {
       case IsHmrcTaxPeriod(el) => el
     }
-    val futureListOfTaxPeriodDes =
-      gformConnector
-        .getAllTaxPeriods(hmrcTaxPeriodIdentifiers)
-        .map(i => i.flatMap(j => j.obligation.obligations.flatMap(h => makeMap(j.id, h.obligationDetails))))
-    futureListOfTaxPeriodDes.map(i => i.toMap)
+    for {
+      currentIdNumber <- authService.evaluateSubmissionReference(
+                 hmrcTaxPeriodIdentifiers match {
```

Comment:
This pattern match is ok, but it should be done as first thing (ie. on line 73) since if the list is empty we know there is no work to do and we can immediately return.